File: exampleProjects/TypescriptExample/src/de/sots/cellarsandcentaurs/domain/service/Creatures.ts
import { Creature } from '../model/Creature';
import { CreatureId } from '../model/CreatureId';
import { NoSuchCreatureException } from '../model/NoSuchCreatureException';

export interface Creatures {
    save(creature: Creature): void;
    find(id: CreatureId): Creature | Promise<Creature>;  // Either returns a Creature or a Promise resolving to Creature
}

File: exampleProjects/PythonExample/src/de/sots/cellarsandcentaurs/domain/services/creatures.py
from abc import ABC, abstractmethod

import de.sots.cellarsandcentaurs.domain.models
import de.sots.cellarsandcentaurs.domain.models.creature as crt
from ..models.no_such_creature_exception import NoSuchCreatureException

class Creatures(ABC):

    @abstractmethod
    def save(self, creature: crt.Creature):
        """Save a creature to the persistence layer."""
        pass

    @abstractmethod
    def find(self, creature_id: de.sots.cellarsandcentaurs.domain.models.CreatureId) -> Creature:
        """Find a creature by its ID.

        Raises:
            NoSuchCreatureException: If the creature is not found.
        """
        pass